enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,284 @@
+  - "@dave-tucker"
+  - "@dcbw"
+  - "@trozet"
+
+approvers:
+  - "@danwinship"
+  - "@knobunc"
+
+api-approvers:
+  - "@danwinship"
+
+creation-date: 2022-06-15
+
+last-updated: 2022-07-19
+
+status: implementable
+
+tracking-link:
+  - https://issues.redhat.com/browse/SDN-2781
+---
+
+# Ingress node Firewall using eBPF
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+OCP customers would like to secure OCP's nodes from external attacks by configuring ingress firewall policies.
+eBPF provides a very flexible filtering mechanism to allow early detection and filtering.
+eXpress Data Path (XDP) is a feature in the Linux kernel which allows users to execute a user-supplied eBPF program when a packet is received on a network interface (NIC). 
+This program is executed in the NIC’s device driver, right after the interrupt processing and before the network stack. 
+The goal of XDP is to offer comparable performance to kernel bypass solutions while working with the existing kernel networking stack.
+A new cluster-scoped object `IngressNodeFirewall` will be used to configure node ingress firewall match rules and per rule statistics will be avaliable.
+Admin can provision multiple instances of this object while ensuring certain execution order.
+For the best performance XDP programs need to attach to interfaces using the following models:
+- Native XDP \
+  The XDP program is loaded by the network card driver as part of its initial receive path. This also requires support from the network card driver.
+- Offloaded XDP \
+  The XDP program loads directly on the NIC, and executes without using the CPU. This also requires support from the network interface device.
+
+## Motivation
+
+Customers would like to secure OCP cluster from external attacks.
+
+### Goals
+
+- Flexible match filter rules for customers to use with a CRD validation to detect and reject invalid configuration.
+- Provide a mechanism for customers to select which nodes via `nodeSelector`.
+- Provide a mechanism to define attachment interfaces for the node ingress firewall objects.  
+- Ingress node firewall is deployed on a fully operational OCP cluster (i.e. Ingress node firewall is deployed as a day 2 operation and is not enabled by default at the cluster install time).
+- fail safe in case of user provisioned invalid rules which can lockup cluster operation.
+
+### Non-Goals
+- Stateful support
+- Flood Attacks
+
+## Proposal
+
+- customers will label nodes where ingress node policy will be deployed on.
+- customers will deploy daemonSet controller on all the selected nodes by applying `IngressNodeFirewallConfig` object. 
```

Comment:
Why is this needed? Why can't CNO (or whatever it is that processes the IngressNodeFirewallConfig object) deploy the daemonset to the correct nodes automatically?

Reply:
why do you need flexibility? It needs to run on exactly the nodes that have firewall rules doesn't it?

Reply:
> this makes the code simpler

But it makes the user's life more complicated. And it makes the feature less secure; you could write firewall rules and think they were in effect and not realize that they were only in effect on some nodes but not others.

Reply:
all objs have status so u can tell I prefer being explicit stable ds before getting into applying rule to make life easier for all. at this point we can see what will be the customer experience feedback since this is now dev preview